dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/queryengine/Engine.scala

```diff
@@ -73,45 +68,61 @@ class Engine(context: EngineContext) {
     sinks.map(sink => ReachableByTask(List(TaskFingerprint(sink, List(), 0)), Vector()))
   }
 
-  /** Submit tasks to a worker pool, solving them in parallel. Upon receiving results for a task, new tasks are
-    * submitted accordingly. Once no more tasks can be created, the list of results is returned.
-    */
   private def solveTasks(tasks: List[ReachableByTask], sources: Set[CfgNode]): Vector[ReachableByResult] = {
-    var completedResults = List[ReachableByResult]()
+    var completedResults = Vector[ReachableByResult]()
 
     def handleSummary(taskSummary: TaskSummary): Unit = {
       val newTasks = taskSummary.followupTasks
       submitTasks(newTasks, sources)
       val newResults = taskSummary.results
-      completedResults ++= newResults
-    }
 
-    def runUntilAllTasksAreSolved(): Unit = {
-      while (numberOfTasksRunning > 0) {
-        Try {
-          completionService.take.get
-        } match {
-          case Success(resultsOfTask) =>
-            numberOfTasksRunning -= 1
-            handleSummary(resultsOfTask)
-          case Failure(exception) =>
-            numberOfTasksRunning -= 1
-            logger.warn(s"SolveTask failed with exception:", exception)
-            exception.printStackTrace()
-        }
+      if (newResults.length == 0) {
+        return
       }
+
+      synchronized(completedResults ++= newResults)
     }
 
+    def submitTasks(tasks: Vector[ReachableByTask], sources: Set[CfgNode]) = {
+      TaskSolver.futuresStartedCounter.incrementAndGet()
+      Future {
+        tasks.foreach(t => {
+          val ts            = new TaskSolver(t, context, sources)
+          val resultsOfTask = ts.call()
+          handleSummary(resultsOfTask)
+        })
+      }.onComplete(_ => {
+        TaskSolver.futuresEndedCounter.incrementAndGet()
+      })
+    }
+
+    val startTimeSec: Long = System.currentTimeMillis / 1000
     submitTasks(tasks.toVector, sources)
-    runUntilAllTasksAreSolved()
-    deduplicate(completedResults.toVector)
-  }
 
-  private def submitTasks(tasks: Vector[ReachableByTask], sources: Set[CfgNode]) = {
-    numberOfTasksRunning += tasks.size
-    tasks.foreach(t => completionService.submit(new TaskSolver(t, context, sources)))
+    do {
```

Comment:
Do I understand correctly that we're waiting for 1 second in a loop here and polling for results? Is there maybe a more elegant way where we block until all calculations are done?

Reply:
We are polling the counters every second to indicate whether all jobs are done. 

A more elegant way could be to use semaphores, condition variables or barriers - depending on which ones are available in Scala. Signal task completion from futures and wake up the main thread for exit after all tasks are done. This will involve some complex logic to achieve sync but I thought that wasn't necessary since this rudimentary way also does the job.

Reply:
Ok. That means though that every `reachableBy` invocation takes at least a second, even if it is a very simple query. I can imagine we'd notice this during interactive use, so, we may not be able to get around building something a bit more evolved. Let me think about it.

Reply:
I have reduced the poll time to 100 milliseconds

Reply:
@fabsx00 

I find that the number of tasks and flows is 70-80% higher on this branch ( with your changes ) compared to `master`

Reply:
@fabsx00 

I have merged latest master into this branch. The result table is now the critical section and is under lock. I see that it has a higher locking time causing threads to get stuck on it. Is there a way to reduce the locking time?

I am referring to

`Engine.scala -> addResultsToMainTable() -> synchronized(mainResultTable.add(parentTask, Vector(r.copy(taskStack = newTaskStack, path = newPath))))`